File: Data Structures & Algorithms/interleaving-string/submission-20.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:

    # i   0   1   2   3           j   0   1   2   3
    #     a   a   a   a               b   b   b   b

    # k   0   1   2   3   4   5   6   7
    #     a   a   b   b   b   b   a   a
        M,N,O= len(s1),len(s2),len(s3)
        if M+N != O:
            return False
        dp = {}
        def dfs(i,j):
            k =i+j
            if k == O:
                return True
            if (i,j) in dp:
                return dp[(i,j)]
            if i<M and s1[i]== s3[k] and dfs(i+1,j):
                return True
            if j<N and s2[j] == s3[k] and dfs(i,j+1):
                return True
            dp[(i,j)] = False
            return False
        return dfs(0,0)




        

File: Data Structures & Algorithms/count-paths/submission-20.java
class Solution {
    Integer[][] grid;
    int ROWS,COLS;
    public int uniquePaths(int m, int n) {
        grid = new Integer[m+1][n+1];
        for(Integer[] row:grid){
            Arrays.fill(row,0);
        }
        grid[m-1][n-1] = 1;
        for(int i = m-1;i>=0;i--){
            for(int j = n-1;j>=0;j--){
                grid[i][j] += grid[i+1][j]+grid[i][j+1];
            }
        }
        return grid[0][0];
    }

    public int dfs(int i,int j){
        if(i==ROWS-1 && j == COLS-1){
            return 1;
        }
        if(i<0 || i==ROWS || j<0 || j==COLS){
            return 0;
        }
        if(grid[i][j] != null){
            return grid[i][j];
        }
        grid[i][j] = dfs(i+1,j) + dfs(i,j+1);
        return grid[i][j];
    }
}
